File: extra-code/stage2_code/my_cpp_pkg/src/param_publisher.cpp
// ============================================================
// Parameter(参数)示例:可动态调节频率和内容的 Publisher
// 演示:declare/get parameter、启动时传参、运行中用 ros2 param set 动态修改,
// 以及两种参数回调:on_set(修改前校验,可否决)和 post_set(修改后生效)
// 运行:ros2 run my_cpp_pkg param_publisher --ros-args -p frequency:=2.0
// 动态修改:ros2 param set /param_publisher frequency 5.0
// ============================================================

#include <chrono>
#include <memory>
#include <stdexcept>
#include <string>

// 参数校验回调需要返回这个消息类型,表示"是否允许本次修改"
#include "rcl_interfaces/msg/set_parameters_result.hpp"
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

class ParamPublisherNode : public rclcpp::Node
{
public:
  ParamPublisherNode()
  : Node("param_publisher")
  {
    // declare_parameter 声明参数并给默认值:未声明的参数外部无法读写
    // 默认值的类型同时决定了参数类型(1.0 是 double,"Hello ROS 2" 是 string)
    this->declare_parameter("frequency", 1.0);
    this->declare_parameter("message", "Hello ROS 2");

    // 读取参数实际值(可能来自命令行 -p 或 YAML 文件,没传就用默认值)
    frequency_ = this->get_parameter("frequency").as_double();
    message_ = this->get_parameter("message").as_string();
    // 启动时的非法值直接抛异常终止:参数回调只拦截"运行中"的修改,管不到启动值
    if (frequency_ <= 0.0) {
      throw std::invalid_argument("frequency must be positive");
    }
    publisher_ = this->create_publisher<std_msgs::msg::String>("chatter", 10);
    // 定时器周期由参数决定:频率 frequency Hz 对应周期 1/frequency 秒
    timer_ = this->create_wall_timer(
      std::chrono::duration<double>(1.0 / frequency_),
      std::bind(&ParamPublisherNode::timer_callback, this));
    // 注册两种参数回调(返回的 handle 必须保存,否则回调立即失效):
    //   on_set  - 参数被修改"之前"调用,用于校验,返回 false 可拒绝本次修改
    //   post_set - 修改已生效"之后"调用,用于应用新值(如重建定时器)
    on_set_callback_handle_ = this->add_on_set_parameters_callback(
      std::bind(&ParamPublisherNode::on_set_parameters, this, std::placeholders::_1));
    post_set_callback_handle_ = this->add_post_set_parameters_callback(
      std::bind(&ParamPublisherNode::post_set_parameters, this, std::placeholders::_1));

    RCLCPP_INFO(
      this->get_logger(), "Param publisher started: frequency=%.2f, message=\"%s\"",
      frequency_, message_.c_str());
  }

private:
  // 定时器回调:按当前参数指定的内容发布消息
  void timer_callback()
  {
    auto msg = std_msgs::msg::String();
    msg.data = message_;
    publisher_->publish(msg);
    RCLCPP_INFO(this->get_logger(), "Publishing: \"%s\"", msg.data.c_str());
  }

  // 修改前校验回调:一次可能同时修改多个参数,所以收到的是 vector
  // 只要发现 frequency 非正数就返回 successful=false,整批修改都会被拒绝
  rcl_interfaces::msg::SetParametersResult on_set_parameters(
    const std::vector<rclcpp::Parameter> & parameters)
  {
    for (const auto & parameter : parameters) {
      if (parameter.get_name() == "frequency") {
        const double value = parameter.as_double();
        if (value <= 0.0) {
          rcl_interfaces::msg::SetParametersResult result;
          result.successful = false;
          result.reason = "frequency must be positive";
          return result;
        }
      }
    }

    // 校验通过:允许修改生效
    rcl_interfaces::msg::SetParametersResult result;
    result.successful = true;
    return result;
  }

  // 修改后生效回调:此时新值已通过校验,把它们同步到成员变量
  void post_set_parameters(const std::vector<rclcpp::Parameter> & parameters)
  {
    bool frequency_changed = false;
    for (const auto & parameter : parameters) {
      if (parameter.get_name() == "frequency") {
        frequency_ = parameter.as_double();
        frequency_changed = true;
      } else if (parameter.get_name() == "message") {
        message_ = parameter.as_string();
      }
    }

    // 定时器创建后周期不可改,只能取消旧定时器再按新频率重建一个
    if (frequency_changed) {
      timer_->cancel();
      timer_ = this->create_wall_timer(
        std::chrono::duration<double>(1.0 / frequency_),
        std::bind(&ParamPublisherNode::timer_callback, this));
    }

    RCLCPP_INFO(
      this->get_logger(), "Parameters updated: frequency=%.2f, message=\"%s\"",
      frequency_, message_.c_str());
  }

  double frequency_;
  std::string message_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr timer_;
  // 两个回调 handle:节点存活期间必须持有,handle 销毁则对应回调自动注销
  rclcpp::node_interfaces::OnSetParametersCallbackHandle::SharedPtr on_set_callback_handle_;
  rclcpp::node_interfaces::PostSetParametersCallbackHandle::SharedPtr post_set_callback_handle_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  // spin 事件循环同时处理定时器回调和外部的参数读写请求
  rclcpp::spin(std::make_shared<ParamPublisherNode>());
  rclcpp::shutdown();
  return 0;
}

File: extra-code/stage2_code/my_py_pkg/my_py_pkg/param_publisher.py
# ============================================================
# 参数（Parameter）系统示例：可配置的 Publisher
# 演示概念：declare_parameter 声明参数、运行中动态改参数、
# 参数校验回调（on_set）与生效回调（post_set）
# 运行方式：ros2 run my_py_pkg param_publisher
# 运行中修改参数示例：
#   ros2 param set /param_publisher frequency 2.0
# ============================================================
import rclpy
# SetParametersResult：参数校验回调必须返回的结果类型
from rcl_interfaces.msg import SetParametersResult
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from rclpy.parameter import Parameter
from std_msgs.msg import String


class ParamPublisher(Node):
    """Publish a configurable String message and react to parameter updates."""

    def __init__(self):
        super().__init__("param_publisher")
        # 声明参数并给默认值；只有声明过的参数才能被
        # ros2 param 命令读写，也可在启动时用 -p 覆盖
        self.declare_parameter("frequency", 1.0)
        self.declare_parameter("message", "Hello ROS 2")

        self.publisher_ = self.create_publisher(String, "chatter", 10)
        # 读取参数当前值，缓存为成员变量供回调使用
        self.frequency_ = float(self.get_parameter("frequency").value)
        self.message_ = str(self.get_parameter("message").value)
        if self.frequency_ <= 0.0:
            raise ValueError("frequency must be positive")
        # 周期 = 1/频率，例如 frequency=2.0 时每 0.5 秒发一次
        self.timer_ = self.create_timer(1.0 / self.frequency_, self.timer_callback)
        # 注册两个参数回调：
        # on_set 在参数生效前触发，用于校验（可拒绝非法值）；
        # post_set 在参数生效后触发，用于应用新值
        self.on_set_callback_handle_ = self.add_on_set_parameters_callback(
            self.on_set_parameters)
        self.post_set_callback_handle_ = self.add_post_set_parameters_callback(
            self.post_set_parameters)
        self.get_logger().info(
            f"Param publisher started: frequency={self.frequency_}, "
            f'message="{self.message_}"')

    def timer_callback(self):
        # 定时发布参数指定的消息内容
        msg = String()
        msg.data = self.message_
        self.publisher_.publish(msg)
        self.get_logger().info(f'Publishing: "{msg.data}"')

    def on_set_parameters(self, params: list[Parameter]):
        # 校验回调：参数生效前调用。返回 successful=False
        # 就会拒绝这次修改（ros2 param set 会报失败）
        for param in params:
            if param.name == "frequency":
                if float(param.value) <= 0.0:
                    return SetParametersResult(
                        successful=False,
                        reason="frequency must be positive",
                    )
        return SetParametersResult(successful=True)

    def post_set_parameters(self, params: list[Parameter]):
        # 生效回调：参数已通过校验并写入后调用，
        # 在这里把新值应用到节点的实际行为上
        frequency_changed = False
        for param in params:
            if param.name == "frequency":
                self.frequency_ = float(param.value)
                frequency_changed = True
            elif param.name == "message":
                self.message_ = str(param.value)

        # Timer 周期创建后不能直接改，
        # 频率变了就取消旧 Timer、按新周期重建一个
        if frequency_changed:
            self.timer_.cancel()
            self.timer_ = self.create_timer(
                1.0 / self.frequency_, self.timer_callback)
        self.get_logger().info(
            f"Parameters updated: frequency={self.frequency_}, "
            f'message="{self.message_}"')


def main(args=None):
    rclpy.init(args=args)
    node = ParamPublisher()
    # spin 阻塞运行；参数修改请求也是在事件循环中被处理的
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == "__main__":
    main()
